fcoe: replace fipvlan with fcoemon (#1085325)

rvykydal opened the pull request:
Same as dracut (#1129888), we are moving to use fcoemon which provides more
robust approach than fipvlan.

Resolves: rhbz#1085325

vpodzime commented on blivet/fcoe.py:
```diff
@@ -110,27 +111,69 @@ def addSan(self, nic, dcb=False, auto_vlan=True):
+            while timeout > 0:
+                rc, out = util.run_program_and_capture_output(
+                    ["lldptool", "-p"])
+                if not rc:
```
I prefer using ``if rc != 0`` in these checks as I always has to think about it otherwise. Unless you're strongly against that, please replace these conditions with ``if rc != 0``.

vpodzime commented on blivet/fcoe.py:
```diff
@@ -110,27 +111,69 @@ def addSan(self, nic, dcb=False, auto_vlan=True):
+                if not rc:
+                    break
+                time.sleep(1)
+                log.info("Retrying to set up dcb for fcoe")
```
Looks like this "wait until utility exits successfully or time out" could be a useful function probably in ``iutil.py`` as it may be useful for other modules too.

rvykydal replied:
Note that the timeout is common for a sequence of utilities run. Do you still think such a function could be useful? (It would have crazy amount of arguments.)

vpodzime replied:
It wouldn't if you could give it something like a ``remaining`` argument that would tell it how much time is still available. But it would have to return the updated value together with other things which would probably make it too complex. So I'm okay with leaving this as it is.

vpodzime commented on blivet/fcoe.py:
```diff
@@ -110,27 +111,69 @@ def addSan(self, nic, dcb=False, auto_vlan=True):
-                rc, out = util.run_program_and_capture_output(["fipvlan", '-c', '-s', '-f',
-                                                   "-fcoe",  nic])
+                udev.settle()
+                time.sleep(3)
```
Is this sleep really needed? And if so, are 3 seconds always going to be enough? I think we at least need some comment here.

rvykydal replied:
Alright, I tracked it back to
https://bugzilla.redhat.com/show_bug.cgi?id=813057#c4
so what about adding
# Sleep for 3 s to allow dcb negotiation (#813057)

vpodzime replied:
Works for me.

vpodzime commented:
Looks good to me otherwise.

bcl commented on blivet/fcoe.py:
```diff
@@ -110,27 +111,69 @@ def addSan(self, nic, dcb=False, auto_vlan=True):
+                    ["lldptool", "-p"])
+                if not rc:
+                    break
+                log.info("Waiting for lldpad to be ready")
```
these are going to clutter up the logs, maybe only log if the timeout is exceeded. while / else could be used for that.

rvykydal replied:
Yes, I am going to what you suggest, please ignore the patch added to this branch, your way is nicer.

rvykydal replied:
Ah, sorry, actually the patch added is the way I'd go, because eg if the timeout appears in 2nd wile loop, the else branches of the loop *and* all subsequent while loops blocks will get executed.

bcl commented on blivet/fcoe.py:
```diff
@@ -110,27 +111,69 @@ def addSan(self, nic, dcb=False, auto_vlan=True):
+                timeout -= 1
+                time.sleep(1)
+
+            while timeout > 0:
```
timeout isn't reset between these blocks.

rvykydal replied:
it is intentional, we want to wait 60 secs maximum for all the tools run (as Chris Leech pointed out when reviewing my former version which had 60s timeout for each tool)

bcl commented on blivet/fcoe.py:
```diff
@@ -145,28 +188,46 @@ def write(self, root):
+    def write_nic_fcoe_cfg(self, nic, dcb=True, auto_vlan=True, enable=True, mode=None, root=""):
+        cfg_dir = root + "/etc/fcoe"
+        example_cfg = os.path.join(cfg_dir, "cfg-ethx")
+        if os.access(example_cfg, os.R_OK):
```
Looks like blivet needs simpleconfig from anaconda.

rvykydal replied:
Yes, I was missing it too.

jkonecny12 commented on blivet/fcoe.py:
```diff
@@ -145,28 +188,46 @@ def write(self, root):
+DCB_REQUIRED="yes"\n
+AUTO_VLAN="yes"\n
+MODE="fabric"\n
+"""
```
Could you please give here better indentation. Maybe concatenation of strings will be better here for readability.

rvykydal commented:
I will, thanks.

rvykydal commented:
Pushed so we can get as early testing as possible.

I addressed these review remarks locally:
- "if not rc" by vpodzime
- comment for time.sleep(3) by vpodzime
- cluttering log messages by bcl
- log message % by clumens
- missing timeout decrement by rvykydal
and tested the patch in beaker.
